### ISSUE TITLE
Thread/Goroutine leak in daemon when writing DAG objects through command line

### ISSUE THREAD

**Issue by iostat** [ipfs/go-ipfs#4256]
#### Version information:
go-ipfs version: 0.4.10-
Repo version: 5
System version: amd64/darwin
Golang version: go1.8.3

#### Machine setup:
MacBook Pro (Retina, 15-inch, Mid 2015)
macOS Sierra 10.12.5 (16F73)
Processor: 2.2 GHz Intel Core i7
Memory: 16 GB 1600 MHz DDR3
Graphics: Intel Iris Pro 1536 MB

#### Description:
The IPFS daemon appears to be leaking threads/goroutines, which causes a panic in the daemon when performing many writes sequentially through the command line.

##### Test setup:
`crash-daemon.sh`
```sh
#!/bin/bash

MAXDATALEN=32
NUMROUNDS=${1:-1024}

function makeRandomData {
  DATASIZE=( $RANDOM % $MAXDATALEN )
  dd if=/dev/urandom bs=$DATASIZE count=1 2>/dev/null | base64
}

for n in `seq 0 $NUMROUNDS`; do 
  echo \"`makeRandomData`\" | ipfs dag put - > /dev/null 2>&1
  if [ "$?" -ne "0" ]; then
    echo crashed after $n rounds
    exit $?
  fi
done

echo ok after $NUMROUNDS rounds
```

##### Test scenario 0 - control group:
```
$ ipfs daemon
# in another window
$ ./crash-daemon.sh 1
ok after 1 rounds
$ ./crash-daemon.sh 1
ok after 1 rounds
$ ./crash-daemon.sh 1
ok after 1 rounds
$ ./crash-daemon.sh 1
ok after 1 rounds
```
This is actually performing two rounds (round 0 and round 1), but does not crash the daemon at all.

##### Test scenario 1 - rapid fire `ipfs dag write`:
```
$ ipfs daemon >crashlog-fast 2>&1
# in another window
$ ./crash-daemon
crashed after 260 rounds
```
In this scenario, I crash almost always in 255-265 rounds when ran immediately after starting the daemon. `crashlog-fast` may be found at this gist: https://gist.github.com/iostat/4fb01464a54e33fefe5b63da7274454d

##### Test scenario 2 - sleep for 5 secs in between volleys of 20 writes:
```
$ ipfs daemon >crashlog-slow 2>&1
# in another window
$ for f in `seq 0 10`; do ./crash-daemon.sh 20; echo 'sleepytime' ; sleep 5; done
ok after 20 rounds
sleepytime
ok after 20 rounds
sleepytime
ok after 20 rounds
sleepytime
crashed after 13 rounds
sleepytime
crashed after 0 rounds
sleepytime
crashed after 0 rounds
sleepytime
crashed after 0 rounds
sleepytime
crashed after 0 rounds
sleepytime
crashed after 0 rounds
sleepytime
crashed after 0 rounds
sleepytime
crashed after 0 rounds
sleepytime
```

In this scenario, I crash almost always after a sum of 70-90 rounds (with 5 secs in between each volley of 20). `crashlog-slow` may be found at this gist: https://gist.github.com/iostat/344abd3f8af8476675cf663a4e5518e2

**Comment by magik6k** [comment]
There were some fixes related to that merged in since 0.4.10. Can you try to reproduce this using 0.4.11-rc2 from http://ipfs.io/ipns/dist.ipfs.io/go-ipfs/v0.4.11-rc2 (or build from current master)?

**Comment by iostat** [comment]
Hey @magik6k !

Looks like 0.4.11-rc2 did the trick! Running steady even up to rapid-fire of 16384, haven't tested past that but it looks like it'll hold up.

Thanks for the quick response! I'll close this now.
